### PR TITLE
Improve startup performance

### PR DESCRIPTION
**Pull request by stsewd**
Some numbers

## Before

```
186.298  000.003: --- NVIM STARTED ---
179.851  000.004: --- NVIM STARTED ---
196.774  000.007: --- NVIM STARTED ---
201.460  000.003: --- NVIM STARTED ---
197.011  000.003: --- NVIM STARTED ---
190.126  000.003: --- NVIM STARTED ---
188.211  000.003: --- NVIM STARTED ---
199.551  000.003: --- NVIM STARTED ---
```

## After

```
172.590  000.004: --- NVIM STARTED ---
176.226  000.004: --- NVIM STARTED ---
167.962  000.003: --- NVIM STARTED ---
170.539  000.003: --- NVIM STARTED ---
170.572  000.004: --- NVIM STARTED ---
170.122  000.005: --- NVIM STARTED ---
173.517  000.004: --- NVIM STARTED ---
176.183  000.003: --- NVIM STARTED ---
```

### CODE REVIEW

**Comment by theHamsta** [comment]
You could also import `nvim-treesitter.install` in `setup` only when it's needed for `ensure_installed`. Maybe we can group small things that could save time in this PR...

**Comment by theHamsta** [comment]
Startup would be even better if we wouldn't enable all modules for all languages I guess. We should only check for is_supported when we actually open a buffer for a filetype for the first time.

**Comment by stsewd** [comment]
> You could also import nvim-treesitter.install in setup only when it's needed for ensure_installed. Maybe we can group small things that could save time in this PR...

Done.

**Review by theHamsta — APPROVED** [review]

**Comment by stsewd** [comment]
I just realized this doesn't make use of query_extensions, but I think they can be merged with base_language_map https://github.com/nvim-treesitter/nvim-treesitter/pull/420 if that PR is merged we don't need to change anything here, otherwise I'll add those too.

**Review by kyazdani42 — APPROVED** [review]
lgtm :)

**Comment by kyazdani42** [comment]
to merge after #420

**Comment by theHamsta** [comment]
Have you tested if it would make any difference to only call get_runtime_files once with `queries/highlights/*.scm` and save the result in a table to look up the parser files?

File: lua/nvim-treesitter/configs.lua [hunk 1]
```diff
@@ -303,7 +303,11 @@ end
 function M.setup(user_data)
   config.modules = vim.tbl_deep_extend('force', config.modules, user_data)
 
-  require'nvim-treesitter.install'.ensure_installed(user_data.ensure_installed)
+  local ensure_installed = user_data.ensure_installed
+  if #ensure_installed > 0 then
```
Comment (theHamsta) [inline lua/nvim-treesitter/configs.lua:307]: this can be `nil`: `if ensure_installed then` or `if ensure_installed and #ensure_installed > 0` then

**Review by theHamsta — CHANGES_REQUESTED** [review]

File: lua/nvim-treesitter/configs.lua (hunk 1 above)
Comment (kyazdani42) [reply lua/nvim-treesitter/configs.lua:307]: `local ensure_install = user_data.ensure_isntalled or {}`

Comment (stsewd) [reply lua/nvim-treesitter/configs.lua:307]: Changed!

**Comment by stsewd** [comment]
> Have you tested if it would make any difference to only call get_runtime_files once with queries/highlights/*.scm and save the result in a table to look up the parser files from the table instead of filesystem?

I didn't, but I did a test always defaulting to `true`, and that brings my startup time to ~130 ms (just in case we want to default to true for languages we know we support).

**Review by theHamsta — APPROVED** [review]

**Pull request merged**